# Sandbox for Learning Cromwell + WDL

## About

This git repo contains a sandbox and experiments for learning and running Cromwell + WDL.

## Installation

### Install Cromwell and Womtools

* First, read the instructions here:  https://cromwell.readthedocs.io/en/stable/tutorials/FiveMinuteIntro
* Then, download the cromwell and womtools JAR files from:  https://github.com/broadinstitute/cromwell/releases/

### Install Everything Else

Download and install bowtie2, samtools, and bcftools:

* http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
* http://www.htslib.org/

Download and install docker:  https://www.docker.com/

(optional) Download and install Jenkins:  https://jenkins.io/

## About the Pipeline

This is a very simple NGS Pipeline that:

* aligns a set of reads to the e-coli genome.
* calls variants via samtools and bcftools.

## Part I:  Run the Pipeline without Cromwell

First, try running the pipeline without Cromwell.  This ensures that you have bowtie2, samtools, and bcftools installed correctly.

* make sure bowtie2, samtools, and bcftools are available in your path.
* run:  `bin/run_ecoli_native.sh`

Within the alignments directory, you should now see a set of output files, including `sim_variants_1.vcf` file.  Verify that the file now exists and that it contains one variant at position 736.

(optional) you can run:  `bin/clean_native.sh` to delete all the intermediate files generated by the pipeline.

## Part II:  Run the WDL Pipeline without Docker

Next, try running the WDL pipeline via Cromwell, but without Docker.

* make sure the at `comwell-xx.jar` and `womtool-xx.jar` are in the root directory of wdl_sandbox.
* run `bin/run_ecoli_local_no_docker.sh`

This script will automatically strip out the docker attributes within the WDL file.

The results of your run will now be somewhere under `cromwell-executions`.

## Part III:  Run the WDL Pipeline with Docker

Next, try running the WDL pipeline via Cromwell, but with Docker.

* as before, make sure the at comwell-xx.jar and womtool-xx.jar are in the root directory of wdl_sandbox.
* make sure you have installed docker.
* run `bin/run_ecoli_local_with_docker.sh`

Wait a few minutes.  This takes longer because of the overhead associated with docker.

The results of your run will now be somewhere under `cromwell-executions`.

## Part IV:  Run the WDL Pipeline via Google Genomics API

[Under Construction]

## Test the WDL Pipeline

[Under Construction]

## Test the WDL Pipeline via Jenkins
